src/Corax/Queries/SortingMatch.cs

```diff
@@ -0,0 +1,325 @@
+                    }
+                    else if (typeof(W) == typeof(NumericalItem<long>))
+                    {
+                        return _comparer.CompareNumerical(((NumericalItem<long>)(object)ix.Value).Value, ((NumericalItem<long>)(object)iy.Value).Value);
+                    }
+                    else if (typeof(W) == typeof(NumericalItem<double>))
+                    {
+                        return _comparer.CompareNumerical(((NumericalItem<double>)(object)ix.Value).Value, ((NumericalItem<double>)(object)iy.Value).Value);
+                    }
+                }
+                else if (readX)
+                {
+                    return 1;
+                }
+
+                return -1;
+            }
+
+            public void Dispose()
+            {
+                ArrayPool<Item>.Shared.Return(_hash);
+            }
+        }
+
+        public SortingMatch(IndexSearcher searcher, in TInner inner, in TComparer comparer, int take = -1)
+        {
+            _searcher = searcher;
+            _inner = inner;
+            _take = take;
+            _comparer = comparer;
+        }
+
+        public long Count => _inner.Count;
+
+        public QueryCountConfidence Confidence => _inner.Confidence;
+
+        public int AndWith(Span<long> prevMatches)
+        {
+            throw new NotSupportedException($"{nameof(SortingMatch<TInner, TComparer>)} does not support the operation of {nameof(AndWith)}.");
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public int Fill(Span<long> matches)
+{
+            return _comparer.FieldType switch
+            {
+                MatchCompareFieldType.Floating => Fill<NumericalItem<double>>(matches),
+                MatchCompareFieldType.Integer => Fill<NumericalItem<long>>(matches),
+                MatchCompareFieldType.Sequence => Fill<SequenceItem>(matches),
+            };
+        }
+
+        [SkipLocalsInit]
+        private int Fill<W>(Span<long> matches) where W : struct
+        {
+            int take = _take <= 0 ? matches.Length : _take;
+
+            // We get the first batch. 
+            int totalMatches = 0;
+            var tmpMatches = matches;
+            while (tmpMatches.Length != 0)
+            {
+                int read = _inner.Fill(tmpMatches);
+                if (read == 0)
+                    break;
+
+                totalMatches += read;
+                tmpMatches = matches.Slice(totalMatches);
+            }
+
+            // We sort the first batch
+            using var hashCacheComparer = new HashCacheMatchComparer<TComparer, W>(_searcher, _comparer);
+            var sorter = new Sorter<long, HashCacheMatchComparer<TComparer, W>>(hashCacheComparer);
+            sorter.Sort(matches.Slice(0, totalMatches));
+
+            Span<long> a = stackalloc long[matches.Length];
+            Span<long> b = stackalloc long[matches.Length];
```

Comment:
I think we shouldn't be cleaning this up, if `Length` is high (16K), that can end up being *very* large.
Better to have an "arena" allocation and use that instead, no?

Reply:
Yes, I plan to use a pattern there that essentially is something like:
```
Span<long> a = matches.Length < 4096 ? stackalloc long[matches.Length] : ArrayPool<long>.Shared.Rent(matches.Length); 
```

Reply:
No, why bother with the pool, just allocate a thread local buffer and get from it

Reply:
Yes, that is an implementation detail. Could be that of basically anything we choose to support that.